pkg/doc/verifiable/credential_jwt.go

```diff
@@ -0,0 +1,260 @@
+	}
+
+	rawClaims := new(jwtRawCredClaims)
+	err = parsedJwt.UnsafeClaimsWithoutVerification(rawClaims)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse JWT claims: %w", err)
+	}
+
+	return rawClaims, nil
+}
+
+// credUnsecuredJWTDecoder parses serialized unsecured JWT.
+type credUnsecuredJWTDecoder struct{}
+
+func (ud *credUnsecuredJWTDecoder) parseJWTClaims(rawJwt []byte) (*JWTCredClaims, error) {
+	_, bytesClaim, err := unmarshalUnsecuredJWT(rawJwt)
+	if err != nil {
+		return nil, fmt.Errorf("failed to decode unsecured JWT")
+	}
+
+	credClaims := new(JWTCredClaims)
+	err = json.Unmarshal(bytesClaim, credClaims)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse JWT claims: %w", err)
+	}
+
+	return credClaims, nil
+}
+
+func (ud *credUnsecuredJWTDecoder) parseJWTRawClaims(rawJwt []byte) (*jwtRawCredClaims, error) {
+	_, bytesClaim, err := unmarshalUnsecuredJWT(rawJwt)
+	if err != nil {
+		return nil, fmt.Errorf("failed to decode unsecured JWT")
+	}
+
+	rawClaims := new(jwtRawCredClaims)
+	err = json.Unmarshal(bytesClaim, rawClaims)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse JWT claims: %w", err)
+	}
+
+	return rawClaims, nil
+}
+
+// decodeCredJWT parses JWT from the specified bytes array in compact format using jwtDecoder.
+// It returns decoded Verifiable Credential refined by JWT Claims in raw byte array and rawCredential form.
+func decodeCredJWT(rawJWT []byte, jwtDecoder credJWTDecoder) ([]byte, *rawCredential, error) {
+	credClaims, err := jwtDecoder.parseJWTClaims(rawJWT)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to decode Verifiable Credential JWT claims: %w", err)
+	}
+
+	rawClaims, err := jwtDecoder.parseJWTRawClaims(rawJWT)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to decode raw Verifiable Credential JWT claims: %w", err)
+	}
+
+	// Apply VC-related claims from JWT.
+	credClaims.refineCredFromJWTClaims()
+	// Complement original "vc" JSON claim with data refined from JWT claims.
+	if err = rawClaims.mergeRefinedVC(credClaims.Credential); err != nil {
+		return nil, nil, fmt.Errorf("failed to merge refined VC: %w", err)
+	}
+
+	var vcData []byte
+	if vcData, err = json.Marshal(rawClaims.Raw); err != nil {
+		return nil, nil, errors.New("failed to marshal 'vc' claim of JWT")
+	}
+
+	return vcData, credClaims.Credential, nil
+}
+
+func (vcc *JWTCredClaims) refineCredFromJWTClaims() {
+	raw := vcc.Credential
+
+	if iss := vcc.Issuer; iss != "" {
+		refineVCIssuerFromJWTClaims(raw, iss)
+	}
+
+	if nbf := vcc.NotBefore; nbf != nil {
+		nbfTime := nbf.Time().UTC()
+		raw.Issued = &nbfTime
+	}
+
+	if jti := vcc.ID; jti != "" {
+		raw.ID = vcc.ID
+	}
+
+	if iat := vcc.IssuedAt; iat != nil {
+		iatTime := iat.Time().UTC()
+		raw.Issued = &iatTime
+	}
+
+	if exp := vcc.Expiry; exp != nil {
+		expTime := exp.Time().UTC()
+		raw.Expired = &expTime
+	}
+}
+
+func decodeCredJWS(rawJwt []byte, fetcher PublicKeyFetcher) ([]byte, *rawCredential, error) {
```

Comment:
Sometimes in your above code, "Signed" is used rather than "JWS".

Reply:
No exact pattern, it's mixed. I will try to improve it.